### PR TITLE
Nested copy support

### LINKED ISSUE #6: Support nested copying
There are several options to implement this (and may not be exclusive)

If one of the properties is also marked with `@MutableCopy` then we can use the mutable type generated instead to allow things like:

```kotlin
data class Person(name: String, job: Job)

data class Job(title: String, department: String)

person.copy { job.department = "QA" }
```

Additionally, or instead, we could generate local mutable entities that we provide access to as nested mutable classes inside the root, `MutablePeople`, in this example.

This could provide similar features as to what optics does (I think).

### PR DESCRIPTION
**Pull request by pablisco**
Closes #6 

Tasks:

*   [x] Make nested copies work with value classes
*   [x] Some clean up
*   [x] Add documentation

### CODE REVIEW

File: ksp/src/main/kotlin/fp/serrano/kopykat/KopyKatProcessor.kt [hunk 1]
```diff
@@ -12,35 +20,33 @@ internal class KopyKatProcessor(
...
-        if (options.mutableCopy) MutableCopyKt.writeTo(codegen)
-      }
-      isDataClass() -> {
-        if (options.copyMap) CopyMapFunctionKt.writeTo(codegen)
-        if (options.mutableCopy) MutableCopyKt.writeTo(codegen)
+        if (options.copyMap) HierarchyCopyMapFunctionKt.writeTo(codegen)
+        if (options.mutableCopy) hierarchyCopyFunctionKt().writeTo(codegen)
       }
-      isValueClass() -> {
-        if (options.valueCopy) ValueCopyFunctionKt.writeTo(codegen)
```
Comment (serras) [inline ksp/src/main/kotlin/fp/serrano/kopykat/KopyKatProcessor.kt:41]: Where has this gone? I guess it's now part of `copyMap`?

File: ksp/src/test/kotlin/fp/serrano/kopykat/NestedMutableCopyTest.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+class NestedMutableCopyTest {
+
+  @Test
+  fun `mutate nested property`() {
+    """
+      |data class Person(val name: String, val job: Job)
+      |data class Job(val title: String)
+      |
+      |val p1 = Person("Alex", Job("Developer"))
+      |val p2 = p1.copy { job.title = "Señor Developer" }
```
Comment (serras) [inline ksp/src/test/kotlin/fp/serrano/kopykat/NestedMutableCopyTest.kt:14]: Qué bonico 😍

**Comment by serras** [comment]
We should also update the docs as part of this PR

File: ksp/src/main/kotlin/fp/serrano/kopykat/KopyKatProcessor.kt (hunk 1 above)
Comment (serras) [inline ksp/src/main/kotlin/fp/serrano/kopykat/KopyKatProcessor.kt:41]: I guess we should remove the `valueCopy` parameter in `KopyKatProvider`

Comment (pablisco) [reply ksp/src/main/kotlin/fp/serrano/kopykat/KopyKatProcessor.kt:41]: Yeah, I managed to merge both Value and Data classes into the same strategy. I tried to make the MutableCopy for value classes also be a value class, but value classes don't allow mutability, duh! :sweat\_smile:

File: ksp/src/test/kotlin/fp/serrano/kopykat/NestedMutableCopyTest.kt (hunk 2 above)
Comment (nomisRev) [reply ksp/src/test/kotlin/fp/serrano/kopykat/NestedMutableCopyTest.kt:14]: Amazing

File: README.md [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
+<!-- TOC -->
+
+* [What can KopyKat do?](#what-can-kopykat-do)
+    * [Mutable `copy`](#mutable-copy)
+        * [Nested mutation](#nested-mutation)
+    * [Mapping `copyMap`](#mapping-copymap)
+    * [Value classes](#value-classes)
+        * [The `copyMap` function](#the-copymap-function)
+        * [The `copy` function](#the-copy-function)
+    * [`copy` for sealed hierarchies](#copy-for-sealed-hierarchies)
```
Comment (serras) [inline README.md:10]: Would you mind moving this one on top of the value classes one? I feel like this is more interesting to most people.

File: README.md [hunk 4]
```diff
@@ -16,15 +25,16 @@ val p1 = Person("Alex", 1)
 val p2 = p1.copy(age = p1.age + 1)  // too many 'age'!
 ```
 
-## The Three Methods
+## What can KopyKat do?
 
-This plug-in generates a couple of new methods that make working with immutable data classes much easier.
+This plug-in generates a couple of new methods that make working with immutable (AKA read only) types more convenient.
```
Comment (serras) [inline README.md:30]:
```suggestion
This plug-in generates a couple of new methods that make working with immutable (read-only) types more convenient.
```

File: README.md [hunk 5]
```diff
@@ -44,6 +54,35 @@ val p3 = p1.copy {
...
+```
+
+Currently, to do mutate inner types you have to do the following:
+
+```kotlin
+val p2 = p1.copy(job = p1.job.copy(title = "Señor Developer"))
+
+```
+
+With KopyKat you can do this in a more readble way:
```
Comment (serras) [inline README.md:76]:
```suggestion
With KopyKat you can do this in a more readable way:
```

File: README.md [hunk 6]
```diff
@@ -90,14 +144,31 @@ This means that the following code works directly, without requiring an intermed
...
+
+Or, you can use a more familiar copy function:
+
+```kotlin
+fun User.takeOver() = copy(name = "Me")
+```
+
+> **Note**
+> KopyKat only generates these if all the subclasses are data or value classes. We can't mutate object types without
+> breaking the world underneath them. And cause a lot of pain.
```
Comment (serras) [inline README.md:161]: ¡Y hasta te has quedado corto con la advertencia!

File: ksp/src/main/kotlin/fp/serrano/kopykat/KopyKatOptions.kt [hunk 7]
```diff
@@ -3,15 +3,13 @@ package fp.serrano.kopykat
...
   val mutableCopy: Boolean,
-  val valueCopy: Boolean,
   val hierarchyCopy: Boolean
 ) {
   companion object {
     fun fromKspOptions(options: Map<String, String>) =
       KopyKatOptions(
         copyMap = options.parseBoolOrTrue("copyMap"),
         mutableCopy = options.parseBoolOrTrue("mutableCopy"),
-        valueCopy = options.parseBoolOrTrue("valueCopy"),
```
Comment (serras) [inline ksp/src/main/kotlin/fp/serrano/kopykat/KopyKatOptions.kt:14]: Are the docs up-to-date with respect to the available options?

**Review by serras — APPROVED** [review]
Looks good! I left a few comments, mostly about the docs.

File: README.md (hunk 3 above)
Comment (pablisco) [reply README.md:10]: Yeah, I was actually contemplating making the value class a note on the relevant section. What do you think?

Comment (serras) [reply README.md:10]: That sounds reasonable. At the end of the day, values classes are the least used of all the things we support.

File: ksp/src/main/kotlin/fp/serrano/kopykat/KopyKatOptions.kt (hunk 7 above)
Comment (pablisco) [reply ksp/src/main/kotlin/fp/serrano/kopykat/KopyKatOptions.kt:14]: No they weren't. Thank you :pray:   
670f29d

File: README.md (hunk 3 above)
Comment (pablisco) [reply README.md:7]: There you go :) fc172ad

**Pull request merged**